armory/data/utils.py

```diff
@@ -72,3 +76,54 @@ def verify_sha256(filepath: str, hash_value: str, block_size: int = 4096):
     value = sha256(filepath, block_size=block_size)
     if value != hash_value:
         raise ValueError(f"sha256 hash of {filepath}: {value} != {hash_value}")
+
+
+def verify_size(filepath: str, file_size: int):
+    size = os.path.getsize(filepath)
+    if size != file_size:
+        raise ValueError(f"file size of {filepath}: {size} != {file_size}")
+
+
+def download_verify_dataset_cache(dataset_dir, checksum_file):
+    with open(checksum_file, "r") as fh:
+        url, file_length, hash = fh.readline().strip().split()
+    # download
+    cache_dir = os.path.join(dataset_dir, "cache")
+    os.makedirs(cache_dir, exist_ok=True)
+    tar_filepath = os.path.join(cache_dir, os.path.basename(url))
+    if not os.path.exists(tar_filepath):
+        curl(url, dataset_dir, tar_filepath)
```

Comment:
Recommend adding a try / except block around the curl (especially for KeyboardInterrupt), and deleting the tar_filepath if it exists after a failure. I was getting weird behavior where if I Ctrl-C to get out of the download the first time, when I run it again it skips to verification, fails, removes the tar file, and then defaults to normal processing. If I Ctrl-C out of *that*, and then run again, it will try to download the cached version again.

I would also add a logging line right before the curl to let the user know you're downloading dataset X.